pymeasure/instruments/lecroy/lecroyT3DSO1204.py

```diff
@@ -0,0 +1,1151 @@
+
+    Attributes:
+        SLEEP_SECONDS: sleep time between commands. If a command is received less than
+        SLEEP_SECONDS after the previous one, the code blocks until at least SLEEP_SECONDS have
+        passed.
+        Because the oscilloscope takes a non neglibile time to perform some operations, it might
+        be needed for the user to tweak the sleep time between commands.
+        The SLEEP_SECONDS is set to 0.5 seconds as default just to be on the safe side. Its
+        optimal value heavily depends on the actual commands and on the connection type,
+        so it is impossible to give a unique value to fit all cases.
+
+    .. code-block:: python
+
+        scope = LeCroyT3DSO1204(resource)
+        scope.autoscale()
+        ch1_data_array, ch1_preamble = scope.download_data(source="C1", points=2000)
+        # ...
+        scope.shutdown()
+    """
+
+    _BOOLS = {True: "ON", False: "OFF"}
+
+    SLEEP_SECONDS = 0.5
+
+    def __init__(self, adapter, **kwargs):
+        super().__init__(adapter, "LeCroy T3DSO1204 Oscilloscope", **kwargs)
+        if hasattr(self.adapter, "connection") and self.adapter.connection is not None:
+            self.adapter.connection.timeout = 7000
+        self._grid_number = 14  # Number of grids in the horizontal direction
+        self._seconds_since_last_write = 0  # Timestamp of the last command
+        self._header_size = 16  # bytes
+        self._footer_size = 2  # bytes
+        self.ch1 = Channel(self, 1)
+        self.ch2 = Channel(self, 2)
+        self.ch3 = Channel(self, 3)
+        self.ch4 = Channel(self, 4)
+        self.waveform_source = "C1"
+        self.default_setup()
+
+    ################
+    # System Setup #
+    ################
+
+    def default_setup(self):
+        """ Setup the oscilloscope for remote operation.
+
+        The COMM_HEADER command controls the
+        way the oscilloscope formats response to queries. This command does not affect the
+        interpretation of messages sent to the oscilloscope. Headers can be sent in their long or
+        short form regardless of the CHDR setting.
+        By setting the COMM_HEADER to OFF, the instrument is going to reply with minimal
+        information, and this makes the response message much easier to parse."""
+        self._comm_header = "OFF"
+
+    def ch(self, source):
+        """ Get channel object from its index or its name. Or if source is "math", just return the
+        scope object.
+
+        :param source: can be 1, 2, 3, 4 or C1, C2, C3, C4, MATH
+        :return: handle to the selected source. """
+        if isinstance(source, str):
+            source = _sanitize_source(source)
+        if source in [1, "C1"]:
+            return self.ch1
+        elif source in [2, "C2"]:
+            return self.ch2
+        elif source in [3, "C3"]:
+            return self.ch3
+        elif source in [4, "C4"]:
+            return self.ch4
+        elif source == "MATH":
+            return self
```

Comment:
why don't you extract the integer from the string and use it to look up the channel in a list?
_sanitize_source could return the channel number (and maybe -1 for math), then you `return self.channels[index]`.

Is it on purpose, that the Math channel is the oscilloscope itself?

Reply:
The _sanitize_source function is used in many places where a string is expected, so I would avoid changing it. In commit bb59bd1a47499ae9022a9f813cf9a2009bd0ef96, I tried to be more concise when selecting the channel object. Would it be acceptable?

Reply:
Yes the math "channel" returns the scope itself because the syntax of the MATH functions is ambiguous, enough that it cannot be completely treated as a standalone channel.

Reply:
We might have some clean-up/better mechanisms for this soon with #718.